File: transfer.py
import json
import time
import asyncio
import base64
from typing import Dict
from arweave import Wallet
import requests

# 定义币对配置
TOKENS = {
    "wAR": "xU9zFkq3X2ZQ6olwNVvr1vUWIjc3kXTWr7xKQD6dh10",
    "qAR": "NG-0lVX882MG5nhARrSzyprEK6ejonHpdUmaaMPsHE8",
    "agent": "8rbAftv7RaPxFjFk5FGUVAVCSjGQB4JHDcb9P9wCVhQ"
}

PAIRS = [
    ("wAR", "qAR"),
    ("wAR", "agent"),
    ("qAR", "agent")
]

class DexClient:
    def __init__(self, wallet_file_path: str):
        self.process_id = "N_JfhIr5Bwz6VTnbL0quOIzn4tgw3P-zxMo0jt6Mk1g"
        # 初始化钱包 - 直接传入文件路径
        self.wallet = Wallet(wallet_file_path)
        
        # AO API endpoint
        self.ao_url = "https://cu2800.ao-testnet.xyz/"  # AO API endpoint

    def sign_message(self, tags: Dict) -> Dict:
        """签名消息"""
        # 创建交易
        transaction = {
            "tags": tags,
            "target": self.process_id,
            "owner": self.wallet.address
        }
        
        # 将交易转换为JSON字符串并签名
        tx_data = json.dumps(transaction)
        signature = self.wallet.sign(tx_data.encode('utf-8'))
        
        signature_b64 = base64.b64encode(signature).decode('utf-8')
        
        # 添加签名到交易
        transaction["signature"] = signature_b64
        
        return transaction

    async def get_result(self, message_id: str) -> Dict:
        """获取消息结果，类似于aoconnect的result函数"""
        try:
            print(f"Getting result for message: {message_id}")
            response = requests.get(
                f"{self.ao_url}/result", 
                params={
                    "message": message_id,
                    "process": self.process_id
                }
            )
            
            print(f"Result response status: {response.status_code}")
            print(f"Result content: {response.text}")
            
            if response.status_code == 200:
                try:
                    result = response.json()
                    return {
                        "Messages": result.get("Messages", []),
                        "Spawns": result.get("Spawns", []),
                        "Output": result.get("Output", {}),
                        "Error": result.get("Error")
                    }
                except json.JSONDecodeError as e:
                    print(f"Failed to parse result as JSON: {response.text}")
                    return None
            else:
                print(f"Failed to get result: {response.text}")
                return None
                
        except Exception as e:
            print(f"Error getting result: {e}")
            return None

    async def send_and_wait(self, tags: Dict) -> Dict:
        """发送消息并等待结果"""
        try:
            # 签名消息
            signed_tx = self.sign_message(tags)
            
            # 打印请求数据用于调试
            print(f"Sending request to AO API:")
            print(f"URL: {self.ao_url}/message")
            print(f"Data: {json.dumps(signed_tx, indent=2)}")
            
            # 发送到AO
            headers = {
                'Content-Type': 'application/json'
            }
            response = requests.post(
                f"{self.ao_url}/message",
                json=signed_tx,
                headers=headers
            )
            
            # 打印响应数据用于调试
            print(f"Response status code: {response.status_code}")
            print(f"Response content: {response.text}")
            
            if response.status_code != 200:
                raise Exception(f"Failed to send message: {response.text}")
            
            try:
                response_data = response.json()
            except json.JSONDecodeError as e:
                print(f"Failed to parse response as JSON: {response.text}")
                raise Exception(f"Invalid JSON response: {e}")
                
            message_id = response_data.get("id")
            if not message_id:
                raise Exception("No message ID in response")
            
            # 等待并获取结果
            max_attempts = 10
            attempt = 0
            
            while attempt < max_attempts:
                result = await self.get_result(message_id)
                if result:
                    if result.get("Error"):
                        raise Exception(f"Transaction failed: {result['Error']}")
                    return result
                
                attempt += 1
                await asyncio.sleep(2)  # 等待2秒后重试
                
            raise Exception("Failed to get transaction result")
            
        except Exception as e:
            print(f"Error in send_and_wait: {e}")
            return None

    async def get_swap_output(self, token: str, quantity: str) -> Dict:
        """获取swap输出预估"""
        try:
            tags = {
                "Action": "Get-Swap-Output",
                "Token": token,
                "Quantity": quantity,
                "Swapper": self.wallet.address
            }
            return await self.send_and_wait(tags)
        except Exception as e:
            print(f"Error in get_swap_output: {e}")
            return None

class ArbitrageBot:
    def __init__(self, wallet_file_path: str):
        self.dex = DexClient(wallet_file_path)
        self.wallet_address = self.dex.wallet.address
        
    async def monitor_prices(self):
        """监控价格变化寻找套利机会"""
        while True:
            try:
                # 检查所有币对的套利机会
                for token1_name, token2_name in PAIRS:
                    token1_id = TOKENS[token1_name]
                    token2_id = TOKENS[token2_name]
                    print(f"\nChecking pair {token1_name}-{token2_name}...")
                    await self.check_arbitrage_opportunity(token1_id, token2_id, token1_name, token2_name)
                
                await asyncio.sleep(10)  # 10秒检查一次
                
            except Exception as e:
                print(f"Error monitoring prices: {e}")
                await asyncio.sleep(5)

    async def check_arbitrage_opportunity(self, token_a: str, token_b: str, token_a_name: str, token_b_name: str):
        """检查是否存在套利机会"""
        test_amount = "1000"  # 测试金额
        
        try:
            # 获取 A->B 的输出预估
            print(f"Checking {token_a_name} -> {token_b_name}...")
            output_a_to_b = await self.dex.get_swap_output(token_a, test_amount)
            
            if output_a_to_b and output_a_to_b.get("Output"):
                output_amount = output_a_to_b["Output"]["tags"]["Output"]
                print(f"{test_amount} {token_a_name} -> {output_amount} {token_b_name}")
                
                # 获取 B->A 的输出预估
                print(f"Checking {token_b_name} -> {token_a_name}...")
                output_b_to_a = await self.dex.get_swap_output(token_b, output_amount)
                
                if output_b_to_a and output_b_to_a.get("Output"):
                    final_amount = int(output_b_to_a["Output"]["tags"]["Output"])
                    initial_amount = int(test_amount)
                    profit = final_amount - initial_amount
                    
                    print(f"{output_amount} {token_b_name} -> {final_amount} {token_a_name}")
                    
                    if profit > 0:
                        print(f"Found arbitrage opportunity in {token_a_name}-{token_b_name}!")
                        print(f"Profit: {profit} {token_a_name} ({(profit/initial_amount)*100:.2f}%)")
                    else:
                        print(f"No arbitrage opportunity found in {token_a_name}-{token_b_name}")
                        print(f"Loss: {profit} {token_a_name} ({(profit/initial_amount)*100:.2f}%)")
                else:
                    print(f"Failed to get {token_b_name}->{token_a_name} output estimate.")
            else:
                print(f"Failed to get {token_a_name}->{token_b_name} output estimate.")
        except Exception as e:
            print(f"Error checking arbitrage opportunity: {e}")

async def main():
    # 初始化套利机器人
    bot = ArbitrageBot(r"E:\blockchain\ARWEAVE\botega\wallet.json")
    
    # 开始监控价格
    await bot.monitor_prices()

if __name__ == "__main__":
    asyncio.run(main())

File: transfer.js
const {
    createDataItemSigner,
    message,
    result
} = require("@permaweb/aoconnect");
const Arweave = require('arweave');
const fs = require('fs');
const path = require('path');

// 初始化 Arweave
const arweave = Arweave.init({
    host: 'arweave.net',
    port: 443,
    protocol: 'https'
});

// 定义钱包文件路径
const WALLET_FILE_PATH = path.join(__dirname, 'wallet.json');
// 确保钱包文件存在
if (!fs.existsSync(WALLET_FILE_PATH)) {
    console.error('Wallet file not found at:', WALLET_FILE_PATH);
    process.exit(1);
}

// 定义币对配置，包含精度信息
const TOKENS = {
    "wAR": {
        id: "xU9zFkq3X2ZQ6olwNVvr1vUWIjc3kXTWr7xKQD6dh10",
        denomination: 12
    },
    "qAR": {
        id: "NG-0lVX882MG5nhARrSzyprEK6ejonHpdUmaaMPsHE8",
        denomination: 12
    },
    "agent": {
        id: "8rbAftv7RaPxFjFk5FGUVAVCSjGQB4JHDcb9P9wCVhQ",
        denomination: 18
    }
};

const PAIRS = [
    ["wAR", "qAR"],
    ["wAR", "agent"],
    ["qAR", "agent"]
];

// 辅助函数：将普通数字转换为链上使用的大数
function toChainValue(amount, decimals) {
    return BigInt(amount) * BigInt(10) ** BigInt(decimals);
}

// 辅助函数：将链上大数转换为普通数字
function fromChainValue(amount, decimals) {
    return Number(amount) / (10 ** decimals);
}

class DexClient {
    constructor(walletFilePath) {
        this.processId = "N_JfhIr5Bwz6VTnbL0quOIzn4tgw3P-zxMo0jt6Mk1g";
        
        try {
            this.wallet = JSON.parse(fs.readFileSync(walletFilePath));
            this.walletAddress = arweave.wallets.jwkToAddress(this.wallet).then(address => {
                console.log('Wallet address:', address);
                return address;
            });
            this.signer = createDataItemSigner(this.wallet);
        } catch (error) {
            console.error('Error initializing DexClient:', error);
            throw error;
        }
    }

    async sendMessage(tags) {
        try {
            const address = await this.walletAddress;
            console.log('Sending message with tags:', tags);

            // 生成唯一的消息ID
            const messageId = `${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;

            const messageData = {
                process: this.processId,
                tags: tags, // 直接使用传入的tags数组
                signer: this.signer,
            };

            console.log('Sending message with data:', messageData);
            const sentMessageId = await message(messageData);
            console.log('Message sent, ID:', sentMessageId);
            
            const res = await result({
                message: sentMessageId,
                process: this.processId
            });
            
            console.log('Result received:', res);
            return res;
        } catch (error) {
            console.error('Error sending message:', error);
            throw error;
        }
    }

    async getReserves() {
        try {
            const tags = [
                { name: "Action", value: "Get-Reserves" }
            ];
            const result = await this.sendMessage(tags);
            if (result && result.Output && result.Output.tags) {
                // 将结果转换为BigInt
                const reserves = {};
                for (const [token, value] of Object.entries(result.Output.tags)) {
                    if (token !== 'Action') {
                        reserves[token] = BigInt(value);
                    }
                }
                return reserves;
            }
            return null;
        } catch (error) {
            console.error('Error in getReserves:', error);
            return null;
        }
    }

    async getSwapOutput(token, quantity) {
        try {
            const address = await this.walletAddress;
            const tags = [
                { name: "Action", value: "Get-Swap-Output" },
                { name: "Token", value: token },
                { name: "Quantity", value: String(quantity) },
                { name: "Swapper", value: address }
            ];

            return await this.sendMessage(tags);
        } catch (error) {
            console.error('Error in getSwapOutput:', error);
            return null;
        }
    }
}

class ArbitrageBot {
    constructor(walletFilePath) {
        this.dex = new DexClient(walletFilePath);
        this.wallet = JSON.parse(fs.readFileSync(walletFilePath));
    }

    async checkArbitrageOpportunity(token1Name, token2Name) {
        try {
            const token1 = TOKENS[token1Name];
            const token2 = TOKENS[token2Name];
            const testAmount = 1000; // 基础单位，将被转换为链上值

            // 获取当前储备金
            const reserves = await this.dex.getReserves();
            if (reserves) {
                console.log('\nCurrent Reserves:');
                console.log(`${token1Name}: ${fromChainValue(reserves[token1.id], token1.denomination)}`);
                console.log(`${token2Name}: ${fromChainValue(reserves[token2.id], token2.denomination)}`);
            }

            // 检查 token1 -> token2
            console.log(`\nChecking ${token1Name} -> ${token2Name}...`);
            const chainAmount = toChainValue(testAmount, token1.denomination);
            const output1to2 = await this.dex.getSwapOutput(token1.id, chainAmount.toString());
            
            if (output1to2 && output1to2.Output) {
                const normalizedOutput1to2 = fromChainValue(BigInt(output1to2.Output), token2.denomination);
                console.log(`${testAmount} ${token1Name} -> ${normalizedOutput1to2} ${token2Name}`);

                // 检查 token2 -> token1 的返回路径
                const chainAmount2 = toChainValue(normalizedOutput1to2, token2.denomination);
                const output2to1 = await this.dex.getSwapOutput(token2.id, chainAmount2.toString());

                if (output2to1 && output2to1.Output) {
                    const finalAmount = fromChainValue(BigInt(output2to1.Output), token1.denomination);
                    const profit = finalAmount - testAmount;
                    const profitPercentage = (profit / testAmount) * 100;

                    console.log(`${normalizedOutput1to2} ${token2Name} -> ${finalAmount} ${token1Name}`);
                    
                    if (profit > 0) {
                        console.log(`Found arbitrage opportunity!`);
                        console.log(`Profit: ${profit} ${token1Name} (${profitPercentage.toFixed(2)}%)`);
                    } else {
                        console.log(`No arbitrage opportunity`);
                        console.log(`Loss: ${profit} ${token1Name} (${profitPercentage.toFixed(2)}%)`);
                    }
                }
            }
        } catch (error) {
            console.error('Error checking arbitrage opportunity:', error);
        }
    }

    async monitorPrices() {
        while (true) {
            try {
                for (const [token1Name, token2Name] of PAIRS) {
                    console.log(`\nChecking pair ${token1Name}-${token2Name}...`);
                    await this.checkArbitrageOpportunity(token1Name, token2Name);
                }
                await new Promise(resolve => setTimeout(resolve, 10000));
            } catch (error) {
                console.error('Error monitoring prices:', error);
                await new Promise(resolve => setTimeout(resolve, 5000));
            }
        }
    }
}

// 主函数
async function main() {
    try {
        const bot = new ArbitrageBot(WALLET_FILE_PATH);
        await bot.monitorPrices();
    } catch (error) {
        console.error('Error in main:', error);
    }
}

main();
